src/Microsoft.AspNetCore.Razor.Tools/MetadataCache.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection.PortableExecutable;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Razor.Tools
+{
+    internal class MetadataCache
+    {
+        // Store 500 entries -- arbitrary number
+        private const int CacheSize = 500;
+        private readonly ConcurrentLruCache<string, MetadataCacheEntry> _metadataCache =
+            new ConcurrentLruCache<string, MetadataCacheEntry>(CacheSize, StringComparer.OrdinalIgnoreCase);
+
+        // For testing purposes only.
+        internal ConcurrentLruCache<string, MetadataCacheEntry> Cache => _metadataCache;
+
+        internal Metadata GetMetadata(string fullPath)
+        {
+            var timestamp = GetFileTimeStamp(fullPath);
+
+            // Check if we have an entry in the dictionary.
+            if (_metadataCache.TryGetValue(fullPath, out var entry))
+            {
+                if (timestamp.HasValue && timestamp.Value == entry.Timestamp)
```

Comment:
Should this allow the cache entry to have a newer timestamp?

Reply:
This block of code would benefit from a code comment.

Reply:
Thought about it. We don't super care if the file has a newer/older timestamp. We just want to get metadata for the file currently on that path. If someone replaced a file with another, there is no point in having the old one in cache regardless of the timestamp.
I'll add a comment.

Reply:
Hmm, looking at this - the check looks kinda incorrect. It always the cache entry if it's up to date, doesn't it?

Reply:
I don't follow. Could you clarify?

Reply:
Nevermind. I was looking at some weird code snippet.